File: vb6-mcp-server/src/analysis/scope.ts
/**
 * Scope Definitions
 *
 * Defines scope hierarchy for the symbol table.
 */

import type { SourceRange } from "./position.js";
import type { ScopeId, SymbolId } from "./types.js";

/**
 * The type of scope
 */
export enum ScopeKind {
  /** Module-level scope (file scope) */
  Module = "Module",
  /** Procedure scope (Sub, Function, Property) */
  Procedure = "Procedure",
  /** With block scope (implicit object reference) */
  WithBlock = "WithBlock",
  /** For loop scope (loop variable) */
  ForLoop = "ForLoop",
  /** For Each loop scope */
  ForEachLoop = "ForEachLoop",
}

/**
 * Check if this scope kind introduces a new variable scope
 * (where local variables can be declared)
 */
export function scopeIsVariableScope(kind: ScopeKind): boolean {
  return kind === ScopeKind.Module || kind === ScopeKind.Procedure;
}

/**
 * Get a display name for the scope kind
 */
export function scopeDisplayName(kind: ScopeKind): string {
  const displayNames: Record<ScopeKind, string> = {
    [ScopeKind.Module]: "Module",
    [ScopeKind.Procedure]: "Procedure",
    [ScopeKind.WithBlock]: "With Block",
    [ScopeKind.ForLoop]: "For Loop",
    [ScopeKind.ForEachLoop]: "For Each Loop",
  };
  return displayNames[kind];
}

/**
 * A scope in the scope hierarchy
 */
export interface Scope {
  /** Unique identifier */
  id: ScopeId;
  /** Scope kind */
  kind: ScopeKind;
  /** Parent scope (undefined for module scope) */
  parent?: ScopeId;
  /** Range covered by this scope */
  range: SourceRange;
  /** Symbols declared in this scope (lowercase name -> symbol id) */
  symbols: Map<string, SymbolId>;
  /** Child scopes (in document order) */
  children: ScopeId[];
  /** For WithBlock: the object expression being referenced */
  withObject?: string;
  /** The symbol that created this scope (for procedure scopes) */
  definingSymbol?: SymbolId;
}

/**
 * Create a new scope
 */
export function createScope(
  id: ScopeId,
  kind: ScopeKind,
  parent: ScopeId | undefined,
  range: SourceRange
): Scope {
  return {
    id,
    kind,
    parent,
    range,
    symbols: new Map(),
    children: [],
    withObject: undefined,
    definingSymbol: undefined,
  };
}

/**
 * Add a symbol to a scope
 */
export function addSymbolToScope(
  scope: Scope,
  name: string,
  symbolId: SymbolId
): void {
  // Use lowercase for case-insensitive lookup
  scope.symbols.set(name.toLowerCase(), symbolId);
}

/**
 * Look up a symbol by name in this scope only (case-insensitive)
 */
export function lookupLocalSymbol(
  scope: Scope,
  name: string
): SymbolId | undefined {
  return scope.symbols.get(name.toLowerCase());
}

/**
 * Get all symbol IDs in this scope
 */
export function getScopeSymbols(scope: Scope): SymbolId[] {
  return Array.from(scope.symbols.values());
}

/**
 * Get the number of symbols in this scope
 */
export function getScopeSymbolCount(scope: Scope): number {
  return scope.symbols.size;
}

/**
 * Check if a symbol name exists in this scope
 */
export function scopeHasSymbol(scope: Scope, name: string): boolean {
  return scope.symbols.has(name.toLowerCase());
}

/**
 * Add a child scope
 */
export function addChildScope(scope: Scope, childId: ScopeId): void {
  scope.children.push(childId);
}

File: src/analysis/scope.rs
//! Scope Definitions
//!
//! Defines scope hierarchy for the symbol table.

use std::collections::HashMap;

use super::position::SourceRange;
use super::symbol::SymbolId;

/// Unique identifier for a scope
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub struct ScopeId(pub u32);

/// The type of scope
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ScopeKind {
    /// Module-level scope (file scope)
    Module,
    /// Procedure scope (Sub, Function, Property)
    Procedure,
    /// With block scope (implicit object reference)
    WithBlock,
    /// For loop scope (loop variable)
    ForLoop,
    /// For Each loop scope
    ForEachLoop,
}

impl ScopeKind {
    /// Check if this scope kind introduces a new variable scope
    /// (where local variables can be declared)
    pub fn is_variable_scope(&self) -> bool {
        matches!(self, ScopeKind::Module | ScopeKind::Procedure)
    }

    /// Get a display name for the scope kind
    pub fn display_name(&self) -> &'static str {
        match self {
            ScopeKind::Module => "Module",
            ScopeKind::Procedure => "Procedure",
            ScopeKind::WithBlock => "With Block",
            ScopeKind::ForLoop => "For Loop",
            ScopeKind::ForEachLoop => "For Each Loop",
        }
    }
}

/// A scope in the scope hierarchy
#[derive(Debug, Clone)]
pub struct Scope {
    /// Unique identifier
    pub id: ScopeId,
    /// Scope kind
    pub kind: ScopeKind,
    /// Parent scope (None for module scope)
    pub parent: Option<ScopeId>,
    /// Range covered by this scope
    pub range: SourceRange,
    /// Symbols declared in this scope (lowercase name -> symbol id)
    symbols: HashMap<String, SymbolId>,
    /// Child scopes (in document order)
    pub children: Vec<ScopeId>,
    /// For WithBlock: the object expression being referenced
    pub with_object: Option<String>,
    /// The symbol that created this scope (for procedure scopes)
    pub defining_symbol: Option<SymbolId>,
}

impl Scope {
    /// Create a new scope
    pub fn new(id: ScopeId, kind: ScopeKind, parent: Option<ScopeId>, range: SourceRange) -> Self {
        Self {
            id,
            kind,
            parent,
            range,
            symbols: HashMap::new(),
            children: Vec::new(),
            with_object: None,
            defining_symbol: None,
        }
    }

    /// Add a symbol to this scope
    pub fn add_symbol(&mut self, name: &str, symbol_id: SymbolId) {
        // Use lowercase for case-insensitive lookup
        self.symbols.insert(name.to_lowercase(), symbol_id);
    }

    /// Look up a symbol by name in this scope only (case-insensitive)
    pub fn lookup_local(&self, name: &str) -> Option<SymbolId> {
        self.symbols.get(&name.to_lowercase()).copied()
    }

    /// Get all symbols declared in this scope
    pub fn symbols(&self) -> impl Iterator<Item = SymbolId> + '_ {
        self.symbols.values().copied()
    }

    /// Get the number of symbols in this scope
    pub fn symbol_count(&self) -> usize {
        self.symbols.len()
    }

    /// Check if a symbol name exists in this scope
    pub fn has_symbol(&self, name: &str) -> bool {
        self.symbols.contains_key(&name.to_lowercase())
    }

    /// Add a child scope
    pub fn add_child(&mut self, child_id: ScopeId) {
        self.children.push(child_id);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::analysis::position::SourcePosition;

    #[test]
    fn test_scope_case_insensitive_lookup() {
        let mut scope = Scope::new(
            ScopeId(0),
            ScopeKind::Module,
            None,
            SourceRange::new(SourcePosition::new(0, 0), SourcePosition::new(100, 0)),
        );

        scope.add_symbol("MyVariable", SymbolId(1));

        assert_eq!(scope.lookup_local("MyVariable"), Some(SymbolId(1)));
        assert_eq!(scope.lookup_local("myvariable"), Some(SymbolId(1)));
        assert_eq!(scope.lookup_local("MYVARIABLE"), Some(SymbolId(1)));
        assert_eq!(scope.lookup_local("NotFound"), None);
    }

    #[test]
    fn test_scope_has_symbol() {
        let mut scope = Scope::new(
            ScopeId(0),
            ScopeKind::Procedure,
            Some(ScopeId(1)),
            SourceRange::default(),
        );

        scope.add_symbol("x", SymbolId(0));

        assert!(scope.has_symbol("x"));
        assert!(scope.has_symbol("X"));
        assert!(!scope.has_symbol("y"));
    }
}
